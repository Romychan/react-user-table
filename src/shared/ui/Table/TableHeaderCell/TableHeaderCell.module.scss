@import '~/app/assets/styles/_typography.scss';

.cell {
  @include text-xs;
  @include text-medium;
  color: var(--gray-600);
  padding: 0.5rem 1rem;
  text-align: left;
  cursor: default;
  white-space: nowrap;
  transition: color 0.2s;
  position: relative;

  &:first-child {
    border-top-left-radius: var(--radius-default);
  }

  &:last-child {
    border-top-right-radius: var(--radius-default);
  }

  &.active {
    color: var(--gray-900);

    & .icon {
      opacity: 1;
      stroke: var(--gray-900);
    }
  }

  &.isSortable {
    cursor: pointer;

    &:hover {
      color: var(--gray-900);

      & .icon {
        opacity: 1;
      }
    }
  }

  &.asc {
  }

  &.desc {
    & .icon {
      transform: rotate(0deg);
    }
  }
}

.icon {
  vertical-align: middle;
  opacity: 0;
  transform: rotate(180deg);
  stroke: var(--gray-900);
  margin-left: 0.25rem;
  transition: 0.2s;
}

.backdrop {
  position: fixed;
  left: 0;
  top: 0;
  height: 100%;
  width: 100%;
  user-select: none;
  cursor: col-resize;
  z-index: 9999;
}

.resizer {
  position: absolute;
  right: 0;
  top: 0;
  cursor: col-resize;
  height: 100%;
  width: 5px;
  z-index: 10;
  display: none;
  transition: 0.2s;

  &.active {
    display: block;
  }

  &:hover,
  &:active {
    background: var(--gray-300);
  }
}
