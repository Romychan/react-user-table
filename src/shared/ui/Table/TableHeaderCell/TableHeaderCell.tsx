import { ReactNode } from 'react';

import { useHorizontalResize } from '../../../lib/hooks';
import { TableOrderValue } from '../types';
import { cl } from '../../../lib/utils';
import { Icon } from '../../Icon';

import styles from './TableHeaderCell.module.scss';

interface ITableHeaderCellProps {
  /** Current sorting type */
  order: TableOrderValue;
  /** The callback that will be called when sorting is selected */
  onClick: () => void;
  /** The content of the component */
  children: ReactNode;
  /**
   * If `true`, the column is selected for sorting
   *
   * @default false
   */
  isActive: boolean;
  /**
   * If `true`, enables sorting for the data in this column
   *
   * @default false
   */
  isSortable?: boolean;
  /**
   * If `true`, enables resizing for this column
   *
   * @default false
   */
  isResizable?: boolean;
  /**
   * Width of the column
   *
   * @default 100
   */
  width?: number;
}

/** This component is used together with the `<TableHeader />` to display column names */
export const TableHeaderCell = ({
  isSortable = false,
  isActive = false,
  isResizable = false,
  order = 'none',
  width = 100,
  onClick,
  children,
}: ITableHeaderCellProps) => {
  const { isDraggable, refContainer, refResizer } =
    useHorizontalResize<HTMLTableCellElement>();

  const isCurrentSortable = isSortable && isActive;
  const isAscSort = isCurrentSortable && order === 'ascending';
  const isDescSort = isCurrentSortable && order === 'descending';
  const isDefaultSort = !isAscSort && !isDescSort && isSortable;

  return (
    <th
      ref={isResizable ? refContainer : null}
      onClick={isSortable ? onClick : undefined}
      style={{ width }}
      className={cl(styles.cell, {
        [styles.active]: isCurrentSortable,
        [styles.isSortable]: isSortable,
        [styles.asc]: isAscSort,
        [styles.desc]: isDescSort,
      })}
      aria-sort={isCurrentSortable ? order : 'none'}
      data-testid="table-header-cell"
    >
      {children}
      {isAscSort || isDescSort || isDefaultSort ? (
        <Icon name="chevron" className={styles.icon} size={14} />
      ) : null}
      <div
        className={cl(styles.resizer, {
          [styles.active]: isResizable,
        })}
        data-testid="table-header-cell-resizer"
        ref={refResizer}
      ></div>
      {isDraggable ? <div className={styles.backdrop}></div> : null}
    </th>
  );
};
