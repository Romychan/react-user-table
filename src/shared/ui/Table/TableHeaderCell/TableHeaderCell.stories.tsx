import type { Meta, StoryObj } from '@storybook/react';

import { TableHeaderCell } from './TableHeaderCell';

const meta = {
  title: 'UI/Table/TableHeaderCell',
  component: TableHeaderCell,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  args: {
    isActive: false,
    order: 'ascending',
    width: 200,
  },
  argTypes: { onClick: { action: 'onClick' } },
  decorators: [
    (Story) => (
      <table>
        <thead>
          <tr>
            <Story />
          </tr>
        </thead>
      </table>
    ),
  ],
} satisfies Meta<typeof TableHeaderCell>;

export default meta;
type Story = StoryObj<typeof meta>;

/** This state is used to default. */
export const Default: Story = {
  args: {
    isSortable: false,
    children: 'Default',
  },
};

/** The `order` property is used to change the sorting option. To sort in ascending order, use `order: 'asc'` */
export const AscSorting: Story = {
  args: {
    isActive: true,
    isSortable: true,
    children: 'Asc Sorting',
  },
};

/** To sort in descending order, use `order: 'desc'` */
export const DescSorting: Story = {
  args: {
    isActive: true,
    isSortable: true,
    order: 'descending',
    children: 'Desc Sorting',
  },
};

/** To resize a cell, use `isResizable: true` */
export const Resizable: Story = {
  args: {
    isResizable: true,
    isActive: true,
    isSortable: true,
    width: 100,
    order: 'descending',
    children: 'Resizable',
  },
};
