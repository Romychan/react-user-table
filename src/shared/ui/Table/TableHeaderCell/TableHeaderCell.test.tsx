import { fireEvent, screen, render } from '@testing-library/react';
import { vi } from 'vitest';

import { TableHeaderCell } from './TableHeaderCell';

describe('TableHeaderCell', () => {
  it('should render correctly', () => {
    const handleClick = vi.fn();

    render(
      <table>
        <thead>
          <tr>
            <TableHeaderCell
              isSortable
              isActive={false}
              order="ascending"
              onClick={handleClick}
            >
              Test
            </TableHeaderCell>
          </tr>
        </thead>
      </table>,
    );
    const cellElement = screen.getByTestId('table-header-cell');
    const containerElement = screen.getByText(/Test/i);
    const iconElement = screen.getByTestId('icon');

    expect(containerElement).toBeInTheDocument();
    expect(cellElement).toBeInTheDocument();
    expect(iconElement).toBeInTheDocument();
  });

  it('should handle click if isSortable', () => {
    const handleClick = vi.fn();

    render(
      <table>
        <thead>
          <tr>
            <TableHeaderCell
              isSortable
              isActive
              order="ascending"
              onClick={handleClick}
            >
              Test
            </TableHeaderCell>
          </tr>
        </thead>
      </table>,
    );
    const cellElement = screen.getByTestId('table-header-cell');

    fireEvent.click(cellElement);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('should not handle click if not isSortable', () => {
    const handleClick = vi.fn();

    render(
      <table>
        <thead>
          <tr>
            <TableHeaderCell
              isSortable={false}
              isActive={false}
              order="ascending"
              onClick={handleClick}
            >
              Test
            </TableHeaderCell>
          </tr>
        </thead>
      </table>,
    );
    const cellElement = screen.getByTestId('table-header-cell');
    const iconElement = screen.queryByTestId('icon');

    expect(iconElement).not.toBeInTheDocument();

    fireEvent.click(cellElement);
    expect(handleClick).toHaveBeenCalledTimes(0);
  });

  it('should show ascending sorting if selected and isSortable', () => {
    const handleClick = vi.fn();

    render(
      <table>
        <thead>
          <tr>
            <TableHeaderCell
              isSortable
              isActive
              order="ascending"
              onClick={handleClick}
            >
              Test
            </TableHeaderCell>
          </tr>
        </thead>
      </table>,
    );
    const cellElement = screen.getByTestId('table-header-cell');
    const iconElement = screen.getByTestId('icon');

    expect(cellElement).toHaveAttribute('aria-sort', 'ascending');
    expect(iconElement).toBeInTheDocument();
  });

  it('should show desc sorting if selected and isSortable', () => {
    const handleClick = vi.fn();

    render(
      <table>
        <thead>
          <tr>
            <TableHeaderCell
              isSortable
              isActive
              order="descending"
              onClick={handleClick}
            >
              Test
            </TableHeaderCell>
          </tr>
        </thead>
      </table>,
    );
    const cellElement = screen.getByTestId('table-header-cell');
    const iconElement = screen.getByTestId('icon');

    expect(cellElement).toHaveAttribute('aria-sort', 'descending');
    expect(iconElement).toBeInTheDocument();
  });

  it('should change the cell size if the isResizable is true', () => {
    const handleClick = vi.fn();
    const CLIENT_X_MOVE = 150;
    const DEFAULT_WIDTH = 100;

    render(
      <table>
        <thead>
          <tr>
            <TableHeaderCell
              isSortable
              isActive
              isResizable
              width={100}
              order="ascending"
              onClick={handleClick}
            >
              Test
            </TableHeaderCell>
          </tr>
        </thead>
      </table>,
    );

    const cellElement = screen.getByTestId('table-header-cell');
    const resizerElement = screen.getByTestId('table-header-cell-resizer');

    fireEvent.mouseDown(resizerElement);
    fireEvent.mouseMove(resizerElement, {
      clientX: CLIENT_X_MOVE,
    });
    fireEvent.mouseUp(resizerElement);

    expect(cellElement).toHaveStyle(
      `width: ${DEFAULT_WIDTH + CLIENT_X_MOVE}px;`,
    );
  });

  it('should not change the cell size if the isResizable is false', () => {
    const handleClick = vi.fn();
    const CLIENT_X_MOVE = 150;
    const DEFAULT_WIDTH = 100;

    render(
      <table>
        <thead>
          <tr>
            <TableHeaderCell
              isSortable
              isActive
              isResizable={false}
              width={100}
              order="ascending"
              onClick={handleClick}
            >
              Test
            </TableHeaderCell>
          </tr>
        </thead>
      </table>,
    );

    const cellElement = screen.getByTestId('table-header-cell');
    const resizerElement = screen.getByTestId('table-header-cell-resizer');

    fireEvent.mouseDown(resizerElement);
    fireEvent.mouseMove(resizerElement, {
      clientX: CLIENT_X_MOVE,
    });
    fireEvent.mouseUp(resizerElement);

    expect(cellElement).toHaveStyle(`width: ${DEFAULT_WIDTH}px;`);
  });

  it('should match snapshot', () => {
    const handleClick = vi.fn();
    const container = render(
      <table>
        <thead>
          <tr>
            <TableHeaderCell
              isSortable
              isActive={false}
              order="ascending"
              onClick={handleClick}
            >
              Test
            </TableHeaderCell>
          </tr>
        </thead>
      </table>,
    );

    expect(container).toMatchSnapshot();
  });
});
