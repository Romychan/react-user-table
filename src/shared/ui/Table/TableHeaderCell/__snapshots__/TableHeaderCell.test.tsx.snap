// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TableHeaderCell > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table>
        <thead>
          <tr>
            <th
              aria-sort="none"
              class="_cell_de2dbc _isSortable_de2dbc"
              data-testid="table-header-cell"
              style="width: 100px;"
            >
              Test
              <svg
                class="icon icon-chevron _icon_de2dbc"
                data-testid="icon"
                height="14"
                width="14"
              >
                <use
                  xlink:href="/images/icons.svg#chevron"
                />
              </svg>
              <div
                class="_resizer_de2dbc"
                data-testid="table-header-cell-resizer"
              />
            </th>
          </tr>
        </thead>
      </table>
    </div>
  </body>,
  "container": <div>
    <table>
      <thead>
        <tr>
          <th
            aria-sort="none"
            class="_cell_de2dbc _isSortable_de2dbc"
            data-testid="table-header-cell"
            style="width: 100px;"
          >
            Test
            <svg
              class="icon icon-chevron _icon_de2dbc"
              data-testid="icon"
              height="14"
              width="14"
            >
              <use
                xlink:href="/images/icons.svg#chevron"
              />
            </svg>
            <div
              class="_resizer_de2dbc"
              data-testid="table-header-cell-resizer"
            />
          </th>
        </tr>
      </thead>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
